### DISCUSSION TITLE
Why are migrations applied twice?

### DISCUSSION THREAD

**Discussion by TechAlchemistry** [prisma/prisma#3789] [Q&A] [answered]
When I applied the init migration, I had some manual changes to the DB: replaced one table with a view.
But now when I try to apply a new migration, prisma wants to create this table (which I replaced), but the current migration is not touching this table. Why is prisma trying to apply the migration twice?
How can I disable this behavior?

**Comment by ryands17** [comment, marked as answer]
@RubtsovAV 👋 
Prisma's `model` always maps to a Table in your database. As you've replaced your table with a view, Prisma thinks that the table doesn't exist and tries to add it again. Prisma [doesn't support views](https://github.com/prisma/prisma/issues/678) at the moment so you would need to use [introspection](https://www.prisma.io/docs/reference/tools-and-interfaces/introspection#overview) instead of Migrate if you want to work with views for now.

**Reply by TechAlchemistry** [reply]
I know Prisma does not support views at the moment, and because I created the view manually. But I need to work with this view through Prisma and it works well until I create a new migration. In this new migration, I see steps.json and it only has the changes I made. 
I expected migrations to be applied consistently. But as it turned out, for some reason, old migrations are launched, although they have already been applied before. 
Why need to verify that all tables from the schema exist if we have a ready-made order of steps to be performed? 
Can I turn off this verify?

**Reply by ryands17** [reply]
Currently it's not possible to disable this behaviour as I had mentioned above that Prisma's model maps to a table so whenever you create a new Migration, it would search check which model doesn't have a table and will create a table for that.

To use views along with other tables, you would need to use a [raw query](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/raw-database-access) for the time being.

**Reply by TechAlchemistry** [reply]
Considering that in migrations it is impossible to add a custom script to change or add data during migration, plus this behavior makes the current migration mechanism extremely inconvenient to work with. Which is depressing. Then you have to make your bike and use other tools.

Thank you for your reply and your time. In any case, the Rrisma is the most good ORM, I will hope that in the future there will be a good tool for migrations.

**Reply by ryands17** [reply]
We are working on a new migration strategy and @albertoperdomo can maybe chime in here as well on how we are working to improve the same situation that is not possible currently.

**Reply by albertoperdomo** [reply]
Hello @RubtsovAV,

I'm with the product team at Prisma, currently focusing on migrations.

I totally understand your concerns and disappointment.

As part of our efforts to deliver a production-ready migrations tool, we are trying to tackle both aspects you've previously addressed in your comments:

* migrations being deterministic - migrate should not try to figure out at runtime what it needs to change in the DB
* no ability to do anything beyond what can be expressed in PSL, i.e. views, native types, stored procedures, triggers, or even copying data around

We are going to be tackling both these problems and soon we will have a version that users can try. If you are interested please let me know and drop an email address or handle on our public Slack so I can reach out to you. :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
